nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-client-service/src/main/java/org/apache/nifi/distributed/cache/client/NettyHandshakeHandler.java

```diff
@@ -0,0 +1,150 @@
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * The {@link io.netty.channel.ChannelHandler} responsible for performing the client handshake with the
+ * distributed cache server.
+ */
+public class NettyHandshakeHandler extends ChannelInboundHandlerAdapter {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * The header bytes used to initiate the server handshake.
+     */
+    private static final byte[] MAGIC_HEADER = new byte[]{'N', 'i', 'F', 'i'};
+
+    /**
+     * The synchronization construct used to signal the client application that the handshake has finished.
+     */
+    private final ChannelPromise promiseHandshakeComplete;
+
+    /**
+     * The version of the protocol negotiated between the client and server.
+     */
+    private final AtomicInteger protocol;
+
+    /**
+     * The coordinator used to broker the version of the distributed cache protocol with the service.
+     */
+    private final VersionNegotiator versionNegotiator;
+
+    /**
+     * Constructor.
+     *
+     * @param channel           the channel to which this {@link io.netty.channel.ChannelHandler} is bound.
+     * @param versionNegotiator coordinator used to broker the version of the distributed cache protocol with the service
+     */
+    public NettyHandshakeHandler(final Channel channel, final VersionNegotiator versionNegotiator) {
+        this.promiseHandshakeComplete = channel.newPromise();
+        this.protocol = new AtomicInteger(0);
+        this.versionNegotiator = versionNegotiator;
+    }
+
+    /**
+     * API providing client application with visibility into the handshake process.  Distributed cache requests
+     * should not be sent using this {@link Channel} until the handshake is complete.
+     */
+    public void waitHandshakeComplete() {
+        promiseHandshakeComplete.awaitUninterruptibly();
+    }
+
+    @Override
+    public void channelActive(final ChannelHandlerContext ctx) throws IOException {
+        final ByteBuf byteBufMagic = Unpooled.wrappedBuffer(MAGIC_HEADER);
+        ctx.write(byteBufMagic);
+        logger.info("Magic header written");
+        final int currentVersion = versionNegotiator.getVersion();
+        final ByteBuf byteBufVersion = Unpooled.wrappedBuffer(new OutboundAdapter().write(currentVersion).toBytes());
+        ctx.writeAndFlush(byteBufVersion);
+        logger.info("Protocol version {} proposed", versionNegotiator.getVersion());
+    }
+
+    @Override
+    public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+        if (promiseHandshakeComplete.isSuccess()) {
+            ctx.fireChannelRead(msg);
+        } else {
+            final ByteBuf byteBuf = (ByteBuf) msg;
+            try {
+                processHandshake(ctx, byteBuf);
+            } catch (IOException | HandshakeException e) {
+                throw new IllegalStateException(e);
```

Comment:
It would be helpful to include a message along with the exception.
```suggestion
                throw new IllegalStateException("Handshake Processing Failed", e);
```